nengo/synapses.py

```diff
@@ -221,123 +223,130 @@ def evaluate(self, frequencies):
         y = np.polyval(self.num, w) / np.polyval(self.den, w)
         return y
 
-    def make_step(self, shape_in, shape_out, dt, rng, y0=None,
-                  dtype=np.float64, method='zoh'):
-        """Returns a `.Step` instance that implements the linear filter."""
-        assert shape_in == shape_out
+    def _get_ss(self, dt):
+        A, B, C, D = tf2ss(self.num, self.den)
 
-        num, den = self.num, self.den
-        if self.analog:
-            num, den, _ = cont2discrete((num, den), dt, method=method)
-            num = num.flatten()
+        # discretize (if len(A) == 0, filter is stateless and already discrete)
+        if self.analog and len(A) > 0:
+            A, B, C, D, _ = cont2discrete((A, B, C, D), dt, method=self.method)
 
-        if den[0] != 1.:
-            raise ValidationError("First element of the denominator must be 1",
-                                  attr='den', obj=self)
-        num = num[1:] if num[0] == 0 else num
-        den = den[1:]  # drop first element (equal to 1)
-        num, den = num.astype(dtype), den.astype(dtype)
+        return A, B, C, D
 
-        output = np.zeros(shape_out, dtype=dtype)
-        if len(num) == 1 and len(den) == 0:
-            return LinearFilter.NoDen(num, den, output)
-        elif len(num) == 1 and len(den) == 1:
-            return LinearFilter.Simple(num, den, output, y0=y0)
-        return LinearFilter.General(num, den, output, y0=y0)
-
-    @staticmethod
-    def _make_zero_step(shape_in, shape_out, dt, rng, y0=None,
-                        dtype=np.float64):
-        output = np.zeros(shape_out, dtype=dtype)
-        if y0 is not None:
-            output[:] = y0
+    def make_step(self, shape_in, shape_out, dt, rng, y0=0, dtype=np.float64):
+        """Returns a `.Step` instance that implements the linear filter."""
+        assert shape_in == shape_out
 
-        return LinearFilter.NoDen(np.array([1.]), np.array([]), output)
+        A, B, C, D = self._get_ss(dt)
+
+        # create state memory variable X
+        X = np.zeros((A.shape[0],) + shape_out, dtype=dtype)
+
+        # initialize X using y0 as steady-state output
+        y0 = np.array(y0, copy=False, ndmin=2)
+        if (y0 == 0).all():
+            # just leave X as zeros in this case, so that this value works
+            # for unstable systems
+            pass
+        elif LinearFilter.OneX.check(A, B, C, D, X):
+            # OneX combines B and C into one scaling value `b`
+            b = B.item() * C.item()
+            X[:] = (b / (1 - A.item())) * y0
+        else:
+            # Solve for u0 (input) given y0 (output), then X given u0
+            assert B.ndim == 1 or B.ndim == 2 and B.shape[1] == 1
+            y0 = np.array(y0, copy=False, ndmin=2)
+            IAB = np.linalg.solve(np.eye(len(A)) - A, B)
+            Q = C.dot(IAB) + D  # multiplier from input to output
```

Comment:
FYI, this is the DC gain. See discrete-time equation here https://www.mathworks.com/help/control/ref/dcgain.html and discussion in #1124, which matches the equation here.

Reply:
Noted in the code.